src/mlpack/core/optimizers/adam/nadamax_update.hpp

```diff
@@ -0,0 +1,170 @@
+/**
+ * @file nadamax_update.hpp
+ * @author Sourabh Varshney
+ *
+ * NadaMax update rule. NadaMax is an optimizer that combines the effect of
+ * Adamax and NAG to the gradient descent to improve its Performance.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_CORE_OPTIMIZERS_ADAM_NADAMAX_UPDATE_HPP
+#define MLPACK_CORE_OPTIMIZERS_ADAM_NADAMAX_UPDATE_HPP
+
+#include <mlpack/prereqs.hpp>
+
+namespace mlpack {
+namespace optimization {
+
+/**
+ * NadaMax is an optimizer that combines the AdaMax and NAG.
+ *
+ * For more information, see the following.
+ *
+ * @code
+ * @techreport{Dozat2015,
+ *   title       = {Incorporating Nesterov momentum into Adam},
+ *   author      = {Timothy Dozat},
+ *   institution = {Stanford University},
+ *   address     = {Stanford},
+ *   year        = {2015},
+ *   url         = {https://openreview.net/pdf?id=OM0jvwB8jIp57ZJjtNEZ}
+ * }
+ * @endcode
+ */
+class NadaMaxUpdate
+{
+ public:
+  /**
+   * Construct the Nadamax update policy with the given parameters.
+   *
+   * @param epsilon The epsilon value used to initialise the squared gradient
+   *        parameter.
+   * @param beta1 The smoothing parameter.
+   * @param beta2 The second moment coefficient
+   * @param scheduleDecay The decay parameter for decay coefficients
+   */
+  NadaMaxUpdate(const double epsilon = 1e-8,
+                const double beta1 = 0.9,
+                const double beta2 = 0.99,
+                const double scheduleDecay = 4e-3):
+      epsilon(epsilon),
+      beta1(beta1),
+      beta2(beta2),
+      scheduleDecay(scheduleDecay),
+      iteration(0),
+      cumBeta1(1)
+  {
+    // Nothing to do.
+  }
+
+  /**
+   * The Initialize method is called by SGD Optimizer method before the start 
+   * of the iteration update process.
+   *
+   * @param rows Number of rows in the gradient matrix.
+   * @param cols Number of columns in the gradient matrix.
+   */
+  void Initialize(const size_t rows, const size_t cols)
+  {
+    m = arma::zeros<arma::mat>(rows, cols);
+    u = arma::zeros<arma::mat>(rows, cols);
+  }
+
+  /**
+   * Update step for NadaMax.
+   *
+   * @param iterate Parameters that minimize the function.
+   * @param stepSize Step size to be used for the given iteration.
+   * @param gradient The gradient matrix.
+   */
+  void Update(arma::mat& iterate,
+              const double stepSize,
+              const arma::mat& gradient)
+  {
+    // Increment the iteration counter variable.
+    ++iteration;
+
+    // And update the iterate.
+    m *= beta1;
+    m += (1 - beta1) * gradient;
+
+    u *= beta2;
+    u = arma::max(u, arma::abs(gradient));
```

Comment:
I think ``u = arma::max(u * beta2, arma::abs(gradient));`` might be easier to read, what you think?

Reply:
Ya, that would be better for reading purpose.